src/java/grl/src/grl/driver/GRL_Driver.java

```diff
@@ -232,7 +262,9 @@ public void run()
 		while (!stop && !_startStopUI.is_stopped() && _lbr.getSafetyState().getEmergencyStopInt()==EmergencyStop.INACTIVE) {
 			message_counter+=1;
 			_currentKUKAiiwaState = udpMan.waitForNextMessage();
-			_previousKUKAiiwaState = udpMan.getPrevMessage();
+			
+			//Some bug in this, just set _previousKUKAiiwaState = _currentKUKAiiwaState before looping/continue?
+			_previousKUKAiiwaState = udpMan.getPrevMessage();  
```

Comment:
Sorry, you do have cancelSmartServo() at #L519. So switching to PauseArm/StopArm/Teach should not have any issue. 

https://github.com/tdinesh/grl/blob/devel_control_modes/include/grl/kuka/KukaJAVAdriver.hpp#L613
does not have mutex locks. Maybe that is the reason for crashes in switching _currentKUKAiiwaState.armControlState().stateType()